### ISSUE TITLE
Load More Stack Frames strange behavior

### ISSUE THREAD

**Issue by edumunoz** [Microsoft/vscode#5364]
- VSCode Version: Version 1.0.0-alpha Commit 6ea1266160aa9eb3ed6e71a5099bec3952090037
- OS Version: Ubuntu 14.04

Steps to Reproduce:

1. Compile https://gist.github.com/edumunoz/6e2ca90d4b5d3dbfa5340576e79ac113 with `g++ -g -std=c++11 -pthread hello.cpp -o hello`
3. Install the [C++ debug adapter](https://marketplace.visualstudio.com/items?itemName=ms-vscode.cpptools)
2. Set a breakpoint in line 17 from the gist
3. Have the main thread expanded in the callstack viewlet
4. Expand another thread and click on a frame for which no source is available
5. The "Load More Stack Frames" button appears for the main thread
6. The "Load More Stack Frames" button is clicked, the callstack for the main thread gets duplicated (two `main()` in the screenshot)

Step 3
![loadmoreframes-1](https://cloud.githubusercontent.com/assets/637952/14575064/49c19a7e-0314-11e6-8863-6b457d261f6e.png)

Step 5
![loadmoreframes-2](https://cloud.githubusercontent.com/assets/637952/14575071/5a5a64f6-0314-11e6-8438-f0e9390a7d53.png)

Step 6
![loadmoreframes-3](https://cloud.githubusercontent.com/assets/637952/14575076/64f29410-0314-11e6-8a2f-a06df05cb8e2.png)

I understand if the debug adapter implements the "Load More Frames" feature, this will work. However the experience seems kind of weird if the debug adapter doesn't implement the feature (like C++ debug adapter as of now).

@isidorn @weinand

**Comment by isidorn** [comment]
@edumunoz that is strange, the Load More Stack Frames action should not appear if an adapter is not implementing it. What does the C++ adapter return for the stackTrace request [total frames](https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/parts/debug/common/debugProtocol.d.ts#L397) attribute. If the adapter does not support this the action [should not be shown](https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/parts/debug/browser/debugViewer.ts#L229) at all.

In short, if the adapter does not support this it should not return the `totalFrames`. If the C++ does not return `totalFrames` attribute and this action still appears then it a bug on the vscode side and I will investigate deeper.

**Comment by edumunoz** [comment]
The C++ extension does not include a `totalFrames` attribute in the `StackTraceResponse`.

Let me know if there is an easy way to turn on logging on the messages sent and received by VSCode to the debug adapter.

**Comment by isidorn** [comment]
Nice catch. The issue was connected with #4623 in that a thread thought it's callstack length was a callStack length of some other thread. Fixed this by always respecting what the debug adapter says about the totalFrames, so if they are not supported this will not be possible. Will not look into the other bug...

**Comment by andrewcrawley** [comment]
Verified this is fixed for cpptools in master.

**Comment by isidorn** [comment]
Adding the verified tag then.
